src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -0,0 +1,236 @@
+        /// <param name="method">Method name for the SaveChanges method.</param>
+        /// <param name="options">Options when saving changes.</param>
+        /// <param name="callback">The user callback.</param>
+        /// <param name="state">The user state object.</param>
+        internal BulkUpdateSaveResult(DataServiceContext context, string method, SaveChangesOptions options, AsyncCallback callback, object state)
+            : base(context, method, null, options, callback, state)
+        {
+            Debug.Assert(Util.IsBulkUpdate(options), "the options must have bulk update flag set");
+            this.bulkUpdateGraph = new BulkUpdateGraph(this.RequestInfo);
+        }
+
+        /// <summary>
+        /// Returns an instance of the <see cref="BulkUpdateGraph"/>
+        /// </summary>
+        internal BulkUpdateGraph BulkUpdateGraph
+        {
+            get { return this.bulkUpdateGraph; }
+        }
+
+        protected override Stream ResponseStream => throw new NotImplementedException();
+
+        /// <summary>
+        /// Syncronous bulk update request.
+        /// </summary>
+        /// <typeparam name="T"> The type of the top-level objects to be deep-updated.</typeparam>
+        /// <param name="objects"> The top-level objects of the type to be deep updated.</param>
+        internal void BulkUpdateRequest<T>(params T[] objects)
+        {
+            BuildDescriptorGraph(this.ChangedEntries, true, objects);
+
+            // This call ensures that the top-level descriptors' EntitySetName is set properly before the request processing begins. 
+            this.bulkUpdateGraph.ComputeEntitySetName();
+
+            ODataRequestMessageWrapper bulkUpdateRequestMessage = this.GenerateBulkUpdateRequest();
+
+            if (bulkUpdateRequestMessage != null)
+            {
+                bulkUpdateRequestMessage.SetRequestStream(bulkUpdateRequestMessage.CachedRequestStream);
+
+                try
+                {
+                    this.batchResponseMessage = this.RequestInfo.GetSynchronousResponse(bulkUpdateRequestMessage, false);
+                }
+                catch (DataServiceTransportException ex)
+                {
+                    InvalidOperationException exception = WebUtil.GetHttpWebResponse(ex, ref this.batchResponseMessage);
+
+                    throw exception;
+                }
+            }
+        }
+
+        /// <summary>
+        /// This method processes all the changed descriptors in the entity tracker.
+        /// It loops through all the descriptors and creates relationships between the descriptors if any.
+        /// </summary>
+        /// <typeparam name="T">The type of the top-level resources we want to do a deep update on.</typeparam>
+        /// <param name="descriptors">The list of descriptors in the entity tracker.</param>
+        /// <param name="IfRootObject">True if the object is a root object, i.e, it was passed by the client, otherwise false.</param>

```

Comment:
```suggestion
        /// <param name="IfRootObject">True if the <paramref name="objects"/> are root or top-level objects, i.e., they were passed directly by the client, otherwise false.</param>

```